### PR TITLE
Added filters for programming models and programs in bench.py

### PR DESCRIPTION
**Pull request by arsho**
Added two filters for `--bench` parameter:
- Program filter:
  - `--includeprogram`: Include a specific program to the benchmark. 
  - `--excludeprogram`: Exclude a specific program from the benchmark. Skipped if `--includeprogram` filter is used.
- Programming models filter: 
  - `--includeppm`: Include a specific programming model to the benchmark.
  - `--excludeppm`: Exclude a specific programming model from the benchmark. Skipped if `--includeppm` filter is used.

Example of using filters:
  - `python rosetta.py --bench --includeprogram assign`
  - `python rosetta.py --bench --excludeppm cuda --excludeppm openmp-parallel --excludeprogram polybench`
  - `python rosetta.py --bench --includeppm openmp-parallel --includeppm serial --excludeprogram polybench`

### CODE REVIEW

**Comment by Meinersbur** [comment]
Consider a more expressive title for the PR.

Consider adding hyphens in the option name: `--include-program`. As we discussed the full option would be something like `--filterincludeprogramsubstr` (where `substr` is the default with the other possibilities `exact`, regex`, possibly `wildcard`) which I think is easier to read with hyphens. We can discuss omitting the `filter` part entirely.

**Comment by arsho** [comment]
@Meinersbur ,
Updated the pull request. Added the regex and exact arguments for both ppm and programs.

**Comment by Meinersbur** [comment]
Could you add tests for each filter to check whether they are functioning correctly?

**Review by Meinersbur — CHANGES_REQUESTED** [review]

**Review by Meinersbur — CHANGES_REQUESTED** [review]
The filter logic does not work when multiple different options are given. Eg.

`--filter-include-program-substr=assign --filter-exclude-ppm-exact=cuda`

should match all assign benchmarks except those written in CUDA. Also consider adding tests for all options.

File: rosetta/src/rosetta/driver.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 from .util.support import *
 from .util import invoke
 from .common import *
+from .rosetta_filter import *
```
[the remark this answers is not in the document]
Comment (arsho) [reply rosetta/src/rosetta/driver.py:19]: I added `rosetta_` as `filter` is a built in function in Python (https://docs.python.org/3/library/functions.html#filter).

File: rosetta/src/rosetta/rosetta_filter.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                        help="Only look into programming models that contain this substring")
+    parser.add_argument('--filter-ppm-exclude', '--excludeppm', action='append', default=[],
+                        help="Only exclude programming models that contain this substring")
+
+
+def match_program_filter(b, args):
+    filter_program_include = args.filter_program_include
+    filter_program_exclude = args.filter_program_exclude
+    basename = b.basename
+    if filter_program_include:
```
[the remark this answers is not in the document]
Comment (arsho) [reply rosetta/src/rosetta/rosetta_filter.py]: This is intensional. For both types of filters, either include or exclude will be invoked. 
For example, the following command will show benchmarks for `idioms.assign` skipping the `--exclude-program-exact` filter:

```
python rosetta.py --bench --include-program-exact idioms.assign --exclude-program-exact idioms.pointwise
```

Adding both include and exclude filter for ppm or program is not logical.

**Comment by arsho** [comment]
> The filter logic does not work when multiple different options are given. Eg.
> 
> `--filter-include-program-substr=assign --filter-exclude-ppm-exact=cuda`
> 
> should match all assign benchmarks except those written in CUDA. Also consider adding tests for all options.

Mixing include and exclude filters for PPM and Program actually working. 

Currently, the following command generates reports for `idioms.assign` program for PPM: `serial`, `openmp-parallel`, and `openmp-task` as expected:
```
python rosetta.py --bench --include-program=assign --exclude-ppm-exact=cuda
```

We can work on renaming the argument signature as suggested in above discussion.

File: README.md [hunk 3]
```diff
@@ -119,3 +119,28 @@ python rosetta.py -h
 python rosetta.py --build
 python rosetta.py --bench
 ```
+Rosetta benchmarks can be executed using two types of filters:
+- Program filter:
+  - `--include-program`: Include a specific program to the benchmark that contains this substring. 
+  - `--include-program-exact`: Include a specific program to the benchmark that exactly matches this string.
```
Comment (Meinersbur) [inline README.md:125]: agratbaetb

**Review by Meinersbur — CHANGES_REQUESTED** [review]

**Comment by Meinersbur** [comment]
Some clarification: A benchmark should pass the filter if
 * (There is no filter-include) OR (the benchmark matches one of the filter-include) AND
 * the benchmark matches none of the filter-exclude

This implies that you cannot handle ppm and program programs in two different functions. The current implementation just `return benchmarks` if no include filter was specified, causing that any exclude filter is ignored. In no case should a filter be just ignored.

Could you please add some unittests testing each of the command-line options? (Not just the single filter what I added to save some time for testing). It doesn't need to use the benchmarks in the benchmark directory, you could create some benchmark objects just for testing the filter.

Here is an example what I imagine test could look like:
```
(apple, green)
(apple, red)
(cherry, red)
(orange, orange)

--include-color-exact=red
(apple, red)
(cherry, red)

--exclude-color-exact=red
(apple, green)
(orange, orange)

--include-color-exact=red --exclude-fruit-exact=cherry
(apple, red)

--exclude-fruit-exact=cherry --include-color-exact=red 
(apple, red)

--include-color-regex=.*n.*
(apple, green)
(orange, orange)

--include-color-regex=.*n.* --include-color-exact=red
(apple, green)
(apple, red)
(cherry, red)
(orange, orange)

--include-color-regex=.*n.* --include-color-exact=red  --exclude-color-exact=green
(apple, red)
(cherry, red)
(orange, orange)

--include-color-regex=.*e.* 
(apple, green)
(apple, red)
(cherry, red)
(orange, orange)

--include-color-exact=red --include-fruit-exact=apple 
(apple, green)
(apple, red)
(cherry, red)

--include-color-exact=red --include-fruit-exact=apple --exclude-color-exact=green
(apple, red)
(cherry, red)
```

**Pull request merged**